lib/fs/casefs.go

```diff
@@ -370,77 +377,87 @@ func (r *defaultRealCaser) realCase(name string) (string, error) {
 		return out, nil
 	}
 
-	r.mut.Lock()
-	defer r.mut.Unlock()
-
-	node := r.root
+	var node *caseNode
 	for _, comp := range strings.Split(name, string(PathSeparator)) {
-		if node.dirNames == nil || node.expires.Before(time.Now()) {
+		// Check cache
+		v, ok := r.cache.Get(out)
+		if ok {
+			node = v.(*caseNode)
+			if node.expires.Before(time.Now()) {
+				r.purgeRecursive(out, node.children)
+				ok = false
+			}
+		}
+
+		if !ok {
 			// Haven't called DirNames yet, or the node has expired
 
-			var err error
-			node.dirNames, err = r.fs.DirNames(out)
+			dirNames, err := r.fs.DirNames(out)
 			if err != nil {
 				return "", err
 			}
 
-			node.dirNamesLower = make([]string, len(node.dirNames))
-			for i, n := range node.dirNames {
-				node.dirNamesLower[i] = UnicodeLowercase(n)
+			num := len(dirNames)
+			node = &caseNode{
+				children:    make(map[string]struct{}, num),
+				lowerToReal: make(map[string]string, num),
+			}
+			lastLower := ""
+			for _, n := range dirNames {
+				node.children[n] = struct{}{}
+				lower := UnicodeLowercase(n)
+				if lower != lastLower {
```

Comment:
something is weird here, lastLower is never set and I'm not sure what it's supposed to do.

Reply:
I forgot to assign inside the if here. It's there because in a case-sensitive fs there can be multiple files with the same lowercase, and previously we assigned the first (plus it's a few less assignements). It doesn't matter in practice because a direct lookup will always work in a case-sensitive fs.